diff --git a/engine/forms.py b/engine/forms.py
index 00eb66c..e91dc37 100644
--- a/engine/forms.py
+++ b/engine/forms.py
@@ -1,6 +1,8 @@
 from django import forms
 from engine.models import User
+from django.core.validators import FileExtensionValidator
 from django.contrib.auth.forms import UserCreationForm
+from django.utils import timezone
 from crispy_forms.helper import FormHelper

 from engine.models import Squadra, Soluzione, Jolly, Consegna
@@ -121,4 +123,12 @@ class ModificaJollyForm(forms.Form):


 class UploadGaraForm(forms.Form):
-    gara = forms.FileField(label="Gara", help_text="Scegli il file JSON contente la gara da importare")
+    gara = forms.FileField(
+        label="Gara", help_text="Scegli il file .journal o .json contente la gara da importare",
+        validators=[FileExtensionValidator(allowed_extensions=["journal", "json"])])
+    nome_gara = forms.CharField(
+        label="Nome della gara (richiesto solo per il formato .journal)",
+        initial="Disfida", required=False)
+    data_gara = forms.DateTimeField(
+        label="Data della gara (richiesto solo per il formato .journal)",
+        initial=timezone.now(), required=False)
diff --git a/engine/templates/gara/upload.html b/engine/templates/gara/upload.html
index de5b066..443a8e3 100644
--- a/engine/templates/gara/upload.html
+++ b/engine/templates/gara/upload.html
@@ -17,6 +17,14 @@
         {{ form.gara|as_crispy_field }}
         </div>
     </div>
+    <div class="form-row">
+        <div class="form-group col-md-6">
+        {{ form.nome_gara|as_crispy_field }}
+        </div>
+        <div class="form-group col-md-6">
+        {{ form.data_gara|as_crispy_field }}
+        </div>
+    </div>

     <button type="submit" class="btn btn-primary">Carica</button>
 </form>
diff --git a/engine/views.py b/engine/views.py
index 14c3d33..0e7cccd 100644
--- a/engine/views.py
+++ b/engine/views.py
@@ -14,7 +14,12 @@ from engine.models import User, Gara, Soluzione, Squadra, Evento, Consegna, Joll
 from engine.forms import SignUpForm, RispostaFormset, SquadraFormset, InserimentoForm,\
     ModificaConsegnaForm, ModificaJollyForm, UploadGaraForm
 from engine.formfields import IntegerMultiField
+from mathrace_interaction.journal_reader import journal_reader
+from mathrace_interaction.filter.strip_mathrace_only_attributes_from_imported_turing import (
+    strip_mathrace_only_attributes_from_imported_turing)

+import datetime
+import io
 import json
 import time
 import logging
@@ -245,7 +250,14 @@ class UploadGaraView(PermissionRequiredMixin, SuccessMessageMixin, FormView):

     def form_valid(self, form):
         gara_file = self.request.FILES['gara']
-        gara_json = json.load(gara_file)
+        if gara_file.name.endswith(".json"):
+            gara_json = json.load(gara_file)
+        elif gara_file.name.endswith(".journal"):
+            race_name = form.cleaned_data.get("nome_gara")
+            race_date = form.cleaned_data.get("data_gara")
+            with journal_reader(io.TextIOWrapper(gara_file, encoding="utf-8")) as journal_stream:
+                gara_json = journal_stream.read(race_name, race_date)
+            strip_mathrace_only_attributes_from_imported_turing(gara_json)
         self.gara = Gara.create_from_dict(gara_json)
         return super().form_valid(form)

