diff --git a/engine/templates/classifiche/unica.html b/engine/templates/classifiche/unica.html
index a2ee9be..3c164ef 100644
--- a/engine/templates/classifiche/unica.html
+++ b/engine/templates/classifiche/unica.html
@@ -24,7 +24,11 @@
         </tr></thead>
         <tbody>
             {% for sq in object.get_squadre_order %}
+            {% if sq.num >= start_pos and sq.num <= end_pos %}
             <tr id="riga-{{sq.num}}">
+            {% else %}
+            <tr id="riga-{{sq.num}}" style="display: none;" >
+            {% endif %}
               <th scope="row" style="white-space:pre" id="pos-{{sq.num}}">1°</th>
               <th scope="row" class="text-truncate squadra" style="white-space:pre" id="nome-{{sq.num}}">{{sq.nome}}</th>
               <th scope="row" style="display: none;" id="num-{{sq.num}}">{{sq.num}}</th>
diff --git a/engine/templates/gara/detail.html b/engine/templates/gara/detail.html
index 831ffc8..e2ecc26 100644
--- a/engine/templates/gara/detail.html
+++ b/engine/templates/gara/detail.html
@@ -62,6 +62,13 @@
     {% endif %}
 </a>
 {% if is_admin %}
+<ul>
+<li>
+<a class="nohover" href="{% url 'engine:classifica-unica' object.pk %}?start_pos=2&end_pos=5" target="_blank">
+    Visualizzazione unica per un range specifico di posizioni (querystring <code>?start_pos=2&end_pos=5</code>)
+</a>
+</li>
+</ul>
 {% include 'gara/link_querystring.html' with engine_name='engine:classifica-unica' prefix='Visualizzazione unica' %}
 {% endif %}
 </li>
diff --git a/engine/views.py b/engine/views.py
index cf0f0f5..39d1a14 100644
--- a/engine/views.py
+++ b/engine/views.py
@@ -635,6 +635,15 @@ class ClassificaBaseView(UserPassesTestMixin, DetailView):
         else:
             qs[variable_name] = None
 
+    def _convert_integer(self, variable_name, qs):
+        variable_value = self.request.GET.get(variable_name, None)
+        if variable_value is None:
+            qs[variable_name] = None
+        elif variable_value.isdecimal():
+            qs[variable_name] = int(variable_value)
+        else:
+            qs[variable_name] = None
+
     def _convert_querystring(self):
         qs = {}
         self._convert_time("race_time", qs)
@@ -698,9 +707,19 @@ class UnicaView(ClassificaBaseView):
     model = Gara
     template_name = "classifiche/unica.html"
 
+    def _convert_querystring(self):
+        qs = super()._convert_querystring()
+        self._convert_integer("start_pos", qs)
+        self._convert_integer("end_pos", qs)
+        return qs
+
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
-        context['soluzioni'] = self.object.soluzioni.all().order_by("problema")
+        if context["start_pos"] is None:
+            context["start_pos"] = 1
+        if context["end_pos"] is None:
+            context["end_pos"] = len(self.object.squadre.all())
+        context["soluzioni"] = self.object.soluzioni.all().order_by("problema")
         return context
 
 class ScorrimentoView(ClassificaBaseView):
