diff --git a/engine/static/engine/css/classifiche.css b/engine/static/engine/css/classifiche.css
index 95834f8..a7e8423 100644
--- a/engine/static/engine/css/classifiche.css
+++ b/engine/static/engine/css/classifiche.css
@@ -79,6 +79,82 @@ tr:nth-child(odd) .right-answer {
     background-color: darkgreen;
 }
 
+@keyframes blink-wrong-even {
+    0% {
+        background-color: #dc3545;
+        color: white;
+    }
+    50% {
+        background-color: transparent;
+        color: #dc3545;
+    }
+    100% {
+        background-color: #dc3545;
+        color: white;
+    }
+}
+
+@keyframes blink-wrong-odd {
+    0% {
+        background-color: firebrick;
+        color: white;
+    }
+    50% {
+        background-color: transparent;
+        color: firebrick;
+    }
+    100% {
+        background-color: firebrick;
+        color: white;
+    }
+}
+
+@keyframes blink-right-even {
+    0% {
+        background-color: forestgreen;
+        color: white;
+    }
+    50% {
+        background-color: transparent;
+        color: forestgreen;
+    }
+    100% {
+        background-color: forestgreen;
+        color: white;
+    }
+}
+
+@keyframes blink-right-odd {
+    0% {
+        background-color: darkgreen;
+        color: white;
+    }
+    50% {
+        background-color: transparent;
+        color: darkgreen;
+    }
+    100% {
+        background-color: darkgreen;
+        color: white;
+    }
+}
+
+tr:nth-child(even) .wrong-answer.blink {
+    animation: blink-wrong-even 1s infinite;
+}
+
+tr:nth-child(odd) .wrong-answer.blink {
+    animation: blink-wrong-odd 1s infinite;
+}
+
+tr:nth-child(even) .right-answer.blink {
+    animation: blink-right-even 1s infinite;
+}
+
+tr:nth-child(odd) .right-answer.blink {
+    animation: blink-right-odd 1s infinite;
+}
+
 td.unica_giuste {
     padding-top: 0.1em;
     padding-bottom: 0.1em;
@@ -95,7 +171,7 @@ td.unica_giuste {
     font-size: 130%;
 }
 
-.my-fa-stack {
+.jolly-fa-stack {
     position: relative;
     margin-left: 0.1em;
     min-width: 1em;
@@ -103,6 +179,17 @@ td.unica_giuste {
     display: inline-block;
 }
 
+.arrow-fa-stack {
+    display: inline-block;
+    width: 100%;
+    text-align: center;
+    vertical-align: middle;
+}
+
+.arrow-fa-stack i {
+    font-size: 100%;
+}
+
 .punteggio_unica {
     position: relative;
     top: 0;
@@ -110,8 +197,8 @@ td.unica_giuste {
     z-index: 2;
 }
 
-td.wrong-answer .my-fa-stack,
-td.right-answer .my-fa-stack {
+td.wrong-answer .jolly-fa-stack,
+td.right-answer .jolly-fa-stack {
     position: absolute;
     bottom: 3px;
     right: 1px;
diff --git a/engine/static/engine/js/client.js b/engine/static/engine/js/client.js
index 62e4def..e6a4312 100644
--- a/engine/static/engine/js/client.js
+++ b/engine/static/engine/js/client.js
@@ -70,6 +70,8 @@ class Gara {
         for (var i in data.consegne) {
             this.add_consegna(data.consegne[i]);
         }
+        this.futuro_consegne_posizioni = [];
+        this.passato_consegne_posizioni = [];
     }
 
     add_jolly(event) {
@@ -97,14 +99,14 @@ class Gara {
     set time(value) {
         var nel_futuro = (value >= this.time);  // necessario memorizzare perchè this.update_events cambia internamente il valore a this.time
         console.log("updating consegne");
-        this.update_events(value, nel_futuro, this.futuro_consegne, this.passato_consegne);
+        this.update_events(value, nel_futuro, this.futuro_consegne, this.passato_consegne, this.futuro_consegne_posizioni, this.passato_consegne_posizioni);
         console.log("updating bonus");
-        this.update_events(value, nel_futuro, this.futuro_bonus, this.passato_bonus);
+        this.update_events(value, nel_futuro, this.futuro_bonus, this.passato_bonus, null, null);
         // Finalmente, setta il tempo della gara
         this._time = value;
     }
 
-    update_events(new_time, nel_futuro, futuro, passato) {
+    update_events(new_time, nel_futuro, futuro, passato, futuro_posizioni, passato_posizioni) {
         // Si sposta al tempo specificato, calcolando gli eventi (consegne e bonus) in mezzo
         if (nel_futuro) {
             // Stiamo andando in avanti
@@ -124,6 +126,17 @@ class Gara {
 
                 passato.push(e);
                 futuro.shift();
+
+                if (passato_posizioni !== null && futuro_posizioni !== null) {
+                    var classifica_e;
+                    if (futuro_posizioni.length > 0) {
+                        classifica_e = futuro_posizioni.shift();
+                    }
+                    else {
+                        classifica_e = this.get_classifica_posizioni(this.classifica);
+                    }
+                    passato_posizioni.push(classifica_e);
+                }
             }
         }
         else {
@@ -143,6 +156,16 @@ class Gara {
 
                 passato.pop();
                 futuro.unshift(e);
+
+                if (passato_posizioni !== null && futuro_posizioni !== null) {
+                    if (passato_posizioni.length > 0) {
+                        classifica_e = passato_posizioni.pop();
+                    }
+                    else {
+                        classifica_e = this.get_classifica_posizioni(this.classifica);
+                    }
+                    futuro_posizioni.unshift(classifica_e);
+                }
             }
         }
     }
@@ -212,6 +235,15 @@ class Gara {
         return ret
     }
 
+    get_classifica_posizioni(classifica) {
+        var posizioni = new Array(classifica.length).fill(null);
+        for (var i = 0; i < classifica.length; i++) {
+            var sq = classifica[i].squadra;
+            posizioni[sq.id - 1] = i + 1;
+        }
+        return posizioni;
+    }
+
     get punti_problemi() {
       var ret = [];
       for (var i in this.problemi) {
@@ -435,16 +467,6 @@ class Squadra {
         pts += this.bonus_manuale;
         return pts
     }
-
-    posizione(classifica) {
-        var pos = 1;
-        for(var i=0; classifica[i].squadra!=this; i++) {
-            if (!classifica[i].squadra.ospite) {
-                pos += 1;
-            }
-        }
-        return pos
-    }
 }
 
 class Consegna {
@@ -482,6 +504,10 @@ class ClassificaClient {
         this.following = following;
         this.autoplay = 0;
         this.recalculating = false;
+        // Impostazione specifica della classifica unica
+        var urlParams = new URLSearchParams(window.location.search);
+        var blink = urlParams.get("blink");
+        this.blink = (blink && !isNaN(blink) && Number.isInteger(parseFloat(blink))) ? parseInt(blink) : 0;
     }
 
     init() {
@@ -565,7 +591,8 @@ class ClassificaClient {
     }
 
     _mostraClassifica() {
-        var classifica = this.gara.classifica
+        var classifica = this.gara.classifica;
+        var classifica_posizioni = this.gara.get_classifica_posizioni(classifica);
         var max = classifica.length > 0 ? classifica[0].pts : 0;
         max = Math.max(max,this.gara.n_prob*10*4);
 
@@ -573,7 +600,7 @@ class ClassificaClient {
         for (var i in classifica) {
             var sq = classifica[i].squadra;
             var pts = classifica[i].pts;
-            var pos = sq.posizione(classifica);
+            var pos = classifica_posizioni[sq.id-1];
             var elapsed = (this.gara.time - this.gara.inizio)/1000;
             if (this.gara.cutoff != null && this.gara.cutoff >= pos && !sq.ospite && elapsed >= 60*20)
                 $("#team-"+sq.id).addClass("cutoff");
@@ -587,16 +614,6 @@ class ClassificaClient {
         for (const sq_id of this.following) {
             $("#team-"+sq_id).addClass("following");
         }
-
-        // if (this.cutoff == null) {
-        //     $("#cutoff").addClass('d-none');
-        // }
-        // else {
-        //     var bRect = $("#team-"+cutoff_team.id)[0].getBoundingClientRect();
-        //     var pos = $("#classifica-container")[0].getBoundingClientRect();
-        //     $("#cutoff").css('left', (bRect.right-pos.left-1)+'px');
-        // }
-        // Spara un evento per comunicare che la pagina si è aggiornata
     }
 
     _mostraPuntiProblemi() {
@@ -682,28 +699,72 @@ class ClassificaClient {
               $("#giuste-"+problema).addClass("progress-bar-light");
           }
       }
-      this._mostraUnicaOScorrimento();
+      // Chiama l'implementazione comune
+      var classifica = this.gara.classifica;
+      var classifica_posizioni = this.gara.get_classifica_posizioni(classifica);
+      this._mostraUnicaOScorrimento(classifica, classifica_posizioni);
+      // Aggiungi lampeggio alla risposta
+      var passato_length = this.gara.passato_consegne.length;
+      var oldest_blink = Math.min(this.blink, passato_length);
+      for (var i = passato_length - oldest_blink; i < passato_length; i++) {
+          var e = this.gara.passato_consegne[i];
+          var sq = e.squadra;
+          var r = e.problema;
+          $("#cell-" + classifica_posizioni[sq.id - 1] + "-" + r.id).addClass("blink");
+      }
+      // Aggiungi frecce per il cambiamento di posizione in classifica
+      if (this.blink > 0) {
+          $("#freccia-head").show();
+          for (var i in classifica) {
+              var riga = parseInt(i)+1;
+              $("#freccia-"+riga).show();
+          }
+          $("#freccia-foot").show();
+      }
+      if (oldest_blink > 0) {
+          var classifica_posizioni_oldest_blink = this.gara.passato_consegne_posizioni[passato_length - oldest_blink];
+          for (var i in classifica) {
+              var sq = classifica[i].squadra;
+              var riga = parseInt(i)+1;
+              var differenza_posizioni = classifica_posizioni_oldest_blink[sq.id-1] - classifica_posizioni[sq.id-1];
+              var freccia;
+              if (differenza_posizioni > 0) {
+                  freccia = ClassificaClient.freccia_su;
+              }
+              else if (differenza_posizioni < 0) {
+                  freccia = ClassificaClient.freccia_giu;
+              }
+              else {
+                  freccia = ClassificaClient.uguale;
+              }
+              $("#freccia-"+riga).html(freccia);
+          }
+      }
+      else {
+          $("#freccia-"+riga).html();
+      }
     }
 
     _mostraScorrimento() {
-      this._mostraUnicaOScorrimento();
+      var classifica = this.gara.classifica;
+      var classifica_posizioni = this.gara.get_classifica_posizioni(classifica);
+      this._mostraUnicaOScorrimento(classifica, classifica_posizioni);
     }
 
-    _mostraUnicaOScorrimento() {
-      var classifica = this.gara.classifica
+    _mostraUnicaOScorrimento(classifica, classifica_posizioni) {
       for (var i in classifica) {
           var sq = classifica[i].squadra;
           var riga = parseInt(i)+1;
           if (sq.ospite) $("#riga-"+riga).addClass("text-muted");
           else $("#riga-"+riga).removeClass("text-muted");
-          $("#pos-"+riga).html(sq.posizione(classifica)+"° ");
+          $("#pos-"+riga).html(classifica_posizioni[sq.id-1]+"° ");
           $("#nome-"+riga).html(sq.nome);
           $("#num-"+riga).html(sq.id);
           $("#punt-"+riga).html(""+sq.punteggio);
           for (var j in sq.risposte) {
               var r = sq.risposte[j];
               var text = "";
-              $("#cell-"+riga+"-"+j).removeClass("wrong-answer right-answer")
+              $("#cell-"+riga+"-"+j).removeClass("wrong-answer right-answer blink")
 
               if (r.risolto) {
                   $("#cell-"+riga+"-"+j).addClass("right-answer");
@@ -744,7 +805,19 @@ class ClassificaClient {
     }
 }
 
-ClassificaClient.stella_jolly = `<span class="my-fa-stack">
+ClassificaClient.stella_jolly = `<span class="jolly-fa-stack">
     <i class="fas fa-star fa-stack-1x fa-inverse" style="color:yellow"></i>
     <i class="far fa-star fa-stack-1x" style="color:black"></i>
 </span>`;
+
+ClassificaClient.freccia_su = `<span class="arrow-fa-stack">
+    <i class="fas fa-arrow-up" style="color:forestgreen"></i>
+</span>`;
+
+ClassificaClient.freccia_giu = `<span class="arrow-fa-stack">
+    <i class="fas fa-arrow-down" style="color:firebrick"></i>
+</span>`;
+
+ClassificaClient.uguale = `<span class="arrow-fa-stack">
+    <i class="fas fa-equals" style="color:#212529"></i>
+</span>`;
diff --git a/engine/templates/classifiche/unica.html b/engine/templates/classifiche/unica.html
index fd1ba57..c0f37bd 100644
--- a/engine/templates/classifiche/unica.html
+++ b/engine/templates/classifiche/unica.html
@@ -14,6 +14,7 @@
     <table class="table table-striped table-bordered table-sm maybe-fixed text-center">
         <thead class="thead-light"><tr>
             <th scope="col" style="width:2.5%">Pos.</th>
+            <th scope="col" style="display: none; width:2.5%" id="freccia-head"></th>
             <th scope="col" style="width:8%">Squadra</th>
             <th scope="col" style="display: none;">Num. squadra</th>
             <th scope="col" style="width:3%">Punt.</th>
@@ -30,6 +31,7 @@
             <tr id="riga-{{sq.num}}" style="display: none;" >
             {% endif %}
               <th scope="row" style="white-space:pre" id="pos-{{sq.num}}">1°</th>
+              <th scope="row" style="display: none;" id="freccia-{{sq.num}}"></th>
               <th scope="row" class="text-truncate squadra" style="white-space:pre" id="nome-{{sq.num}}">{{sq.nome}}</th>
               <th scope="row" style="display: none;" id="num-{{sq.num}}">{{sq.num}}</th>
               <th scope="row" style="white-space:pre" id="punt-{{sq.num}}"></th>
@@ -42,6 +44,7 @@
         </tbody>
         <thead class="thead-light"><tr>
             <th scope="row"></th>
+            <th scope="row" style="display: none;" id="freccia-foot"></th>
             <th scope="row"></th>
             <th scope="row" style="display: none;"></th>
             <th scope="row"></th>
diff --git a/engine/templates/gara/detail.html b/engine/templates/gara/detail.html
index 448ff9e..d22469f 100644
--- a/engine/templates/gara/detail.html
+++ b/engine/templates/gara/detail.html
@@ -44,6 +44,11 @@
     Visualizzazione unica per un range specifico di posizioni (querystring <code>?start_pos=2&end_pos=5</code>)
 </a>
 </li>
+<li>
+<a class="nohover" href="{% url 'engine:classifica-unica' object.pk %}?blink=6" target="_blank">
+    Visualizzazione unica con lampeggi per ultime risposte e frecce per cambio posizione (querystring <code>?blink=6</code>)
+</a>
+</li>
 </ul>
 {% include 'gara/link_querystring.html' with engine_name='engine:classifica-unica' prefix='Visualizzazione unica' %}
 {% endif %}
